Mathlib/Topology/EMetricSpace/Arclength.lean

```diff
@@ -0,0 +1,373 @@
+
+theorem arclength_sum {n : ℕ} {u : ℕ → α} (hu : Monotone u) :
+    ∑ i ∈ Finset.range n, arclength f (u i) (u (i + 1)) = arclength f (u 0) (u n) := by
+  induction n with
+  | zero => rw [arclength_self, Finset.sum_range_zero]
+  | succ k ih => rw [Finset.sum_range_succ, ih, arclength_add f (hu <| zero_le k) (hu k.le_succ)]
+
+theorem arclength_sub₀ (hba : b ≤ a) : arclength f a b = arclength f a c - arclength f b c := by
+  rw [arclength_eq_zero f hba, eq_comm]
+  exact tsub_eq_zero_of_le (arclength_anti f c hba)
+
+theorem arclength_sub' (hbc : b ≤ c) (hac : arclength f b c ≠ ∞) :
+    arclength f a b =  arclength f a c - arclength f b c := by
+  rcases le_total a b with (hab | hba)
+  · exact ENNReal.eq_sub_of_add_eq hac (arclength_add f hab hbc)
+  · exact arclength_sub₀ f hba
+
+theorem arclength_sub (hbc : b ≤ c) (hac : arclength f a c ≠ ∞) :
+    arclength f a b = arclength f a c - arclength f b c := by
+  rcases le_total a b with (hab | hba)
+  · exact arclength_sub' f hbc <| ne_top_of_le_ne_top hac <| arclength_anti f c hab
+  · exact arclength_sub₀ f hba
+
+open OrderDual
+
+@[simp]
+theorem arclength_comp_ofDual :
+    arclength (f ∘ ofDual) (toDual b) (toDual a) = arclength f a b := by
+  have : Set.Icc (toDual b) (toDual a) = ofDual ⁻¹' (Set.Icc a b) := by simp
+  simp_rw [arclength, this, eVariationOn.comp_ofDual]
+
+theorem arclength_left_eq :
+    (arclength f · b) = arclength (f ∘ ofDual) (toDual b) ∘ toDual := by
+  ext; simp
+
+theorem arclength_right_eq :
+    arclength f a = (arclength (f ∘ ofDual) · (toDual a)) ∘ toDual := by
+  ext; simp
+
+theorem arclength_Icc_extend (h : a ≤ b) (f : Set.Icc a b → E) :
+    arclength (Set.IccExtend h f) a b = eVariationOn f Set.univ :=
+  (eVariationOn.comp_eq_of_monotoneOn _ _ ((Set.monotone_projIcc h).monotoneOn _)).trans <| by
+    rw [(Set.projIcc_surjOn h).image_eq_of_mapsTo (Set.mapsTo_univ ..)]
+
+section
+
+/-! ### Continuity (in various forms) "around" a point -/
+
+variable [TopologicalSpace α]
+
+theorem continuousOn_Iic_arclength_of_ge (h : b ≤ a) :
+    ContinuousOn (arclength f a) (Set.Iic b) :=
+  continuousOn_const.congr fun _x hx ↦ arclength_eq_zero _ (trans hx h)
+
+theorem continuousOn_Ici_arclength_of_ge (h : b ≤ a) :
+    ContinuousOn (arclength f · b) (Set.Ici a) :=
+  continuousOn_const.congr fun _x hx ↦ arclength_eq_zero _ (trans h hx)
+
+variable [OrderTopology α]
+
+theorem continuousAt_arclength_right (h : b < a) : ContinuousAt (arclength f a) b :=
+  continuousAt_const.congr <| Filter.eventuallyEq_of_mem (Iio_mem_nhds h)
+    fun x hx ↦ (arclength_eq_zero _ (le_of_lt <| by simpa)).symm
+
+theorem continuousAt_arclength_left (h : b < a) : ContinuousAt (arclength f · b) a :=
+  continuousAt_const.congr <| Filter.eventuallyEq_of_mem (Ioi_mem_nhds h)
+    fun x hx ↦ (arclength_eq_zero _ (le_of_lt <| by simpa)).symm
+
+variable (hab : a < b) (hrect : arclength f a b ≠ ∞) /- f is rectifiable on [a,b] -/
```

Comment:
This links it to the `BoundedVariationOn` API, but they are definitionally the same thing.
```suggestion
variable (hab : a < b) (hrect : BoundedVariationOn f (Icc a b)) /- f is rectifiable on [a,b] -/
```